pkg/koordlet/runtimehooks/hooks/terwayqos/terwayqos.go

```diff
@@ -0,0 +1,289 @@
+	"k8s.io/klog/v2"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	slov1alpha1 "github.com/koordinator-sh/koordinator/apis/slo/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/resourceexecutor"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/runtimehooks/hooks"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/runtimehooks/rule"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/statesinformer"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/util"
+)
+
+const (
+	rootPath   = "/var/lib/terway/qos"
+	podConfig  = "pod.json"
+	nodeConfig = "node.json"
+)
+
+const (
+	name        = "TerwayQoS"
+	description = "network qos management"
+
+	ruleNameForNodeQoS = name + " (nodeQoS)"
+	ruleNameForAllPods = name + " (allPods)"
+)
+
+type Plugin struct {
+	executor resourceexecutor.ResourceUpdateExecutor
+
+	lock sync.RWMutex
+	node *Node
+}
+
+func (p *Plugin) Register(op hooks.Options) {
+	klog.V(5).Infof("register hook %v", "terwqy qos configure generator")
+	rule.Register(ruleNameForNodeQoS, description,
+		rule.WithParseFunc(statesinformer.RegisterTypeNodeSLOSpec, p.parseRuleForNodeSLO),
+		rule.WithUpdateCallback(p.update))
+	rule.Register(ruleNameForAllPods, description,
+		rule.WithParseFunc(statesinformer.RegisterTypeAllPods, p.parseForAllPods),
+		rule.WithUpdateCallback(p.update))
+
+	p.executor = op.Executor
+}
+
+func (p *Plugin) parseRuleForNodeSLO(mergedNodeSLOIf interface{}) (bool, error) {
+	mergedNodeSLO := mergedNodeSLOIf.(*slov1alpha1.NodeSLOSpec)
+
+	n := &Node{}
+	err := parseNetQoS(mergedNodeSLO, n)
+	if err != nil {
+		return false, err
+	}
+
+	p.lock.Lock()
+	p.node = n
+	p.lock.Unlock()
+	return true, nil
+}
+
+func (p *Plugin) parseForAllPods(e interface{}) (bool, error) {
+	_, ok := e.(*struct{})
+	if !ok {
+		return false, fmt.Errorf("invalid rule type %T", e)
+	}
+
+	return true, nil
+}
+
+func (p *Plugin) update(target *statesinformer.CallbackTarget) error {
+	if target == nil {
+		return fmt.Errorf("callback target is nil")
+	}
+
+	err := os.MkdirAll(rootPath, os.ModeDir)
+	if err != nil {
+		return err
+	}
+
+	pods := make([]Pod, 0, len(target.Pods))
+	for _, meta := range target.Pods {
+		if meta.Pod == nil {
+			continue
+		}
+
+		v4, v6 := util.GetIPs(meta.Pod)
+
+		ing, egress, err := getPodQoS(meta.Pod.Annotations)
+		if err != nil {
+			klog.Errorf("get pod qos failed, err: %v", err)
+			continue
+		}
+		pods = append(pods, Pod{
+			PodName:      meta.Pod.Name,
+			PodNamespace: meta.Pod.Namespace,
+			PodUID:       string(meta.Pod.UID),
+			Prio:         prioMapping[meta.Pod.Labels[extension.LabelPodQoS]],
+			IPv4:         v4,
+			IPv6:         v6,
+			HostNetwork:  meta.Pod.Spec.HostNetwork,
+			CgroupDir:    meta.CgroupDir,
+			QoSConfig: QoSConfig{
+				IngressBandwidth: ing,
+				EgressBandwidth:  egress,
+			},
+		})
+	}
+	outPods, err := json.Marshal(pods)
+	if err != nil {
+		return err
+	}
+
+	err = os.WriteFile(filepath.Join(rootPath, podConfig), outPods, 0644)
```

Comment:
should use ResourceUpdateExecutor when updating config, which has cache an compare before update content